File: leet-code/234. Palindrome Linkded List/234_palindrome_linked_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# 입력으로 주어지는 노드는 단일 연결 리스트
def is_palindrome(head: ListNode):
    two_step = head
    one_step = head
    rev = None

    while two_step and two_step.next:
        two_step = two_step.next.next
        rev, rev.next, one_step = one_step, rev, one_step.next

    if two_step:
        one_step = one_step.next

    while one_step and one_step.val == rev.val:
        one_step, rev = one_step.next, rev.next

    return not rev


node = ListNode(1)
node.next = ListNode(2)
node.next.next = ListNode(3)
node.next.next.next = ListNode(4)
print(is_palindrome(node))

# 참고
# 파이썬 알고리즘 인터뷰

File: leet-code/234. Palindrome Linkded List/234_palindrome_linked_list.dart
class ListNode {
  int val;
  ListNode? next;
  ListNode([this.val = 0, this.next]);
}

class Solution {
  bool isPalindrome(ListNode? head) {
    // fast 가 첫번째 while 문에서 null 이 될 수 있어서 nullable 로 선언
    ListNode? fast = head!;
    // slow 가 두번째 while 문에서 null 이 될 수 있어서 nullable 로 선언
    ListNode? slow = head;
    // 처음에 null 로 값을 두기 위해 nullable 로 선언
    ListNode? rev = null;

    // ignore: unnecessary_null_comparison
    while (fast != null && fast.next != null) {
      fast = fast.next!.next;
      
      ListNode? tempSlow = slow;
      ListNode? tempRev = rev;
      slow = slow!.next;
      rev = tempSlow;
      rev!.next = tempRev;
    }

    // ignore: unnecessary_null_comparison
    if (fast != null) {
      slow = slow!.next;
    }

    while (slow != null && slow.val == rev!.val) {
      slow = slow.next;
      rev = rev.next;
    }

    return rev == null;
  }
}

void main() {
  ListNode node = ListNode(1);
  node.next = ListNode(2);
  node.next!.next = ListNode(3);
  node.next!.next!.next = ListNode(2);
  node.next!.next!.next!.next = ListNode(1);

  Solution solution = Solution();
  print(solution.isPalindrome(node)); 
}
